cmd/isp/src/protocol.rs

```diff
@@ -0,0 +1,544 @@
+
+        Ok(CommandPacket { packet: v, params: args })
+    }
+
+    fn to_bytes(&self) -> Result<Vec<u8>> {
+        let mut v = Vec::new();
+
+        v.extend_from_slice(&self.packet.as_bytes());
+
+        for e in self.params.iter() {
+            v.extend_from_slice(&e.to_le_bytes());
+        }
+
+        Ok(v)
+    }
+}
+
+pub struct DataPacket {
+    packet: FramingPacket,
+    data: Vec<u8>,
+}
+
+impl DataPacket {
+    fn new_data(args: Vec<u8>) -> Result<DataPacket> {
+        let arg_len: u16 = args.len() as u16;
+
+        let mut f = FramingPacket::new(PacketType::Data);
+
+        f.length_low = (arg_len & 0xFF) as u8;
+        f.length_high = ((arg_len >> 8) & 0xff) as u8;
+
+        let mut crc = CRCu16::crc16xmodem();
+
+        let bytes = f.as_bytes();
+
+        crc.digest(&bytes[..0x4]);
+        crc.digest(&bytes[0x6..]);
+        crc.digest(&args);
+
+        let digest = crc.get_crc();
+
+        f.crc16_low = (digest & 0xff) as u8;
+        f.crc16_high = ((digest >> 8) & 0xff) as u8;
+
+        Ok(DataPacket { packet: f, data: args })
+    }
+
+    fn to_bytes(&self) -> Result<Vec<u8>> {
+        let mut v = Vec::new();
+
+        v.extend_from_slice(&self.packet.as_bytes());
+        v.extend_from_slice(&self.data);
+
+        Ok(v)
+    }
+}
+
+pub fn process_ping(port: &mut dyn serialport::SerialPort) -> Result<()> {
+    let ping = PacketHeader::new(PacketType::Ping);
+
+    port.write_all(&ping.as_bytes())?;
+
+    port.flush()?;
+
+    let mut response_bytes: [u8; 10] = [0; 10];
+
+    port.read_exact(&mut response_bytes)?;
+
+    let response =
+        PingResponse::read_from(&response_bytes[..]).ok_or(anyhow!("ugh"))?;
+
+    if response.header.packet_type != (PacketType::PingResponse as u8) {
+        return Err(anyhow!(
+            "Incorrect ACK byte from ping {:x}",
+            response.header.packet_type
+        ));
+    }
+
+    Ok(())
+}
+
+fn send_ack(port: &mut dyn serialport::SerialPort) -> Result<()> {
+    let packet = PacketHeader::new(PacketType::Ack);
+    port.write_all(&packet.as_bytes())?;
+    port.flush()?;
+
+    Ok(())
+}
+
+fn read_ack(port: &mut dyn serialport::SerialPort) -> Result<()> {
+    let mut ack_bytes: [u8; 2] = [0; 2];
+
+    port.read_exact(&mut ack_bytes)?;
+
+    let ack = PacketHeader::read_from(ack_bytes.as_slice())
+        .ok_or(anyhow!("parse err"))?;
+
+    // Ack abort comes with a response packet explaining why
+    if ack.packet_type == (PacketType::AckAbort as u8) {
+        match read_response(port, ResponseCode::Generic) {
+            Ok(p) => {
+                if p.is_empty() {
+                    return Err(anyhow!(
+                        "Response returned an unknown error code?"
+                    ));
+                }
+                // The return value is always the first parameter
+                let retval = p[0];
+
+                // Some more specific error messages.
+                if retval == 10203 {
```

Comment:
Same as my [suggestion below](https://github.com/oxidecomputer/humility/pull/148/files#r881810637), consider a `match` statement here to make it easier to extend in the future.